### PR TITLE
Transition to BSER v2.

### PR DESCRIPTION
**Pull request by dhruvsinghal**
As per https://github.com/facebook/watchman/wiki/Better-Unicode-handling-plan.

- Added tests specific to BSER v1 (to make sure things don't break during the transition).
- Renamed `STRING` to `BYTESTRING`.
- Added context type for switching BSER versions.
- Added BSER version switching logic to server.
- Added BSER v2 as a capability testable with the `version` command.
- Added logic for negotiating up to BSER v2 if both client and server support it.

### CODE REVIEW

**Comment by facebook-github-bot-2** [comment]
@dhruvsinghal updated the pull request.

File: bser.c [hunk 1]
```diff
@@ -48,7 +48,7 @@ static int bser_real(double val, json_dump_callback_t dump, void *data)
   return dump((char*)&val, sizeof(val), data);
 }
 
-bool bunser_string(const char *buf, json_int_t avail, json_int_t *needed,
+bool bunser_bytestring(const char *buf, json_int_t avail, json_int_t *needed,
```
Comment (sid0) [inline bser.c]: (This should actually be a static function. Feel free to change it to one.)

File: bser.c [hunk 2]
```diff
@@ -70,6 +73,16 @@ bool bunser_bytestring(const char *buf, json_int_t avail, json_int_t *needed,
   return true;
 }
 
+bool bunser_bytestring(const char *buf, json_int_t avail, json_int_t *needed,
```
Comment (sid0) [inline bser.c]: Make this and the next one static functions.

File: bser.c [hunk 3]
```diff
@@ -273,8 +295,7 @@ static int bser_array(const json_t *array,
   return 0;
 }
 
-static int bser_object(json_t *obj,
-    json_dump_callback_t dump, void *data)
+static int bser_object(json_t *obj, json_dump_callback_t dump, void *data)
```
Comment (sid0) [inline bser.c]: This is a whitespace change -- could you make it in a separate commit?

File: bser.c [hunk 4]
```diff
@@ -186,6 +201,13 @@ static int bser_bytestring(const char *str, json_dump_callback_t dump, void *dat
   return 0;
 }
 
+static int bser_bytestring(const char *str, json_dump_callback_t dump,
```
Comment (sid0) [inline bser.c]: I guess you'll introduce a `bser_utf8string` at some point?

File: bser.c [hunk 5]
```diff
@@ -29,11 +29,13 @@
 #define BSER_NULL      0x0a
 #define BSER_TEMPLATE  0x0b
 #define BSER_SKIP      0x0c
+#define BSER_UNICODE   0x0d
```
Comment (sid0) [inline bser.c]: To be consistent with everything else call it `BSER_UTF8`.

**Comment by sid0** [comment]
So something to think about is making sure we have the testing story straight for old server/new client and new server/old client. It's probably not something that you'll have to deal with right now but it'll become important very soon.

File: bser.c (hunk 4 above)
Comment (dhruvsinghal) [inline bser.c]: Yes. I originally wrote bser_/bunser_ versions for utf8 but they are not currently used so the compiler complained about them.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by wez** [comment]
I took a brief look over this, and it is looking good!

The CI failures appear to be due to unconditionally emitting the new UTF8 string encoding; I think you'll need to add a way to explicitly enable that while defaulting to disabled to make them pass and then I think you're about ready to land this initial series--so long as it doesn't change or break the default behavior it's ok to land partially done work towards the larger goal.

File: bser.c [hunk 6]
```diff
@@ -48,7 +50,8 @@ static int bser_real(double val, json_dump_callback_t dump, void *data)
   return dump((char*)&val, sizeof(val), data);
 }
 
-bool bunser_string(const char *buf, json_int_t avail, json_int_t *needed,
+/* Common codepath for bytestring and UTF-8 encoded strings. */
+static bool bunser_string(const char *buf, json_int_t avail, json_int_t *needed,
```
Comment (wez) [inline bser.c]: you may want to rename this or change the type signature to help ensure that you got all references to this name and replaced them with bunser_bytestring

Comment (sid0) [inline bser.c]: This is actually fine -- in d44395242985339e85112bede3c02bca8c63b04b @dhruvsinghal renamed `bunser_string` to `bunser_bytestring`.

File: python/pywatchman/bser.c [hunk 7]
```diff
@@ -53,6 +53,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define BSER_NULL      0x0a
 #define BSER_TEMPLATE  0x0b
 #define BSER_SKIP      0x0c
+#define BSER_UTF8STRING 0x0d
```
Comment (sid0) [inline python/pywatchman/bser.c]: Move this out into a separate diff.

File: bser.c [hunk 8]
```diff
@@ -29,11 +29,13 @@
 #define BSER_NULL      0x0a
 #define BSER_TEMPLATE  0x0b
 #define BSER_SKIP      0x0c
+#define BSER_UTF8STRING 0x0d
```
Comment (sid0) [inline bser.c]: Split it out into a separate diff.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by dhruvsinghal** [comment]
@sid0 Can you have a look sometime and say if the refactor involving the context variable is on the right track? (Note that this build won't pass the tests yet since ASAN found bugs)

File: bser.c [hunk 9]
```diff
@@ -342,23 +374,29 @@ static int measure(const char *buffer, size_t size, void *ptr)
   return 0;
 }
 
-int w_bser_write_pdu(json_t *json, json_dump_callback_t dump, void *data)
+int w_bser_write_pdu(bser_ctx_t *ctx, json_t *json, void *data)
 {
   json_int_t m_size = 0;
+  bser_ctx_t measurement_ctx = *ctx;
+  ctx->dump = measure;
```
Comment (sid0) [inline bser.c]: `measurement_ctx->dump = measure`.

File: bser.c [hunk 10]
```diff
@@ -342,23 +374,29 @@ static int measure(const char *buffer, size_t size, void *ptr)
   return 0;
 }
 
-int w_bser_write_pdu(json_t *json, json_dump_callback_t dump, void *data)
+int w_bser_write_pdu(bser_ctx_t *ctx, json_t *json, void *data)
```
Comment (sid0) [inline bser.c]: Make this `const bser_ctx_t *` everywhere.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

File: python/pywatchman/__init__.py [hunk 11]
```diff
@@ -40,10 +40,10 @@
...
 # Sometimes it's really hard to get Python extensions to compile,
 # so fall back to a pure Python implementation.
-try:
-    from . import bser
-except ImportError:
-    from . import pybser as bser
+#try:
+#    from . import bser
+#except ImportError:
+from . import pybser as bser
```
Comment (sid0) [inline python/pywatchman/__init__.py]: Could you revert these lines of code?

File: python/pywatchman/__init__.py [hunk 12]
```diff
@@ -521,7 +521,60 @@ class ImmutableBserCodec(BserCodec):
...
+        return bser.loads(response, False) # Defaults to BSER v1
+
+
+class Bser2WithFallbackCodec(BserCodec):
+    """ use BSER v2 encoding """
+
+    def __init__(self, transport):
+        super(Bser2WithFallbackCodec, self).__init__(transport)
+        if compat.PYTHON3:
+            self.send(["version", {"required": ["bser-v2"]}])
```
Comment (sid0) [inline python/pywatchman/__init__.py]: Nice :)

File: python/pywatchman/__init__.py [hunk 13]
```diff
@@ -521,7 +521,60 @@ class ImmutableBserCodec(BserCodec):
...
+        else:
+            self.send(["version", {"optional": ["bser-v2"]}])
+
+        capabilities = self.receive()
+
+        if u'error' in capabilities:
```
Comment (sid0) [inline python/pywatchman/__init__.py]: This should be a native string, so just `'error'`

File: python/pywatchman/__init__.py [hunk 14]
```diff
@@ -521,7 +521,60 @@ class ImmutableBserCodec(BserCodec):
...
+          raise Exception('Unsupported BSER version')
+
+        if capabilities[u'capabilities'][u'bser-v2']:
```
Comment (sid0) [inline python/pywatchman/__init__.py]: Both of these should be native strings.

File: python/pywatchman/__init__.py [hunk 15]
```diff
@@ -521,7 +521,60 @@ class ImmutableBserCodec(BserCodec):
...
+            self.bser_version = 2
+        else:
+            self.bser_version = 1
+
+        assert self.bser_version == 2 # For the sake of testing, let fallback
```
Comment (sid0) [inline python/pywatchman/__init__.py]: Could you elaborate about what you're doing with this? These sorts of assertions should live in tests, not in code.

File: python/pywatchman/__init__.py [hunk 16]
```diff
@@ -600,9 +653,9 @@ def __init__(self,
             raise WatchmanError('invalid transport %s' % transport)
 
         sendEncoding = str(sendEncoding or os.getenv('WATCHMAN_ENCODING') or
-                           'bser')
+                           'bser-v2-with-fallback')
```
Comment (sid0) [inline python/pywatchman/__init__.py]: I'd just call it `bser`, to be honest. I'd make the old BSER version `bser-v1`.

File: python/pywatchman/bser.c [hunk 17]
```diff
@@ -971,19 +978,21 @@ static PyObject *bser_loads(PyObject *self, PyObject *args, PyObject *kw)
   int datalen = 0;
   const char *end;
   int64_t expected_len;
+  int bser_version = 1;
   PyObject *mutable_obj = NULL;
   const char *value_encoding = NULL;
   const char *value_errors = NULL;
-  unser_ctx_t ctx = {1, 0};
+  unser_ctx_t ctx = { .mutable = 1, .value_encoding = NULL };
```
Comment (sid0) [inline python/pywatchman/bser.c]: I'm a bit concerned about the `.mutable` stuff and its support on Windows in particular. What does the AppVeyor thing say? Otherwise, just do

```
unser_ctx_t ctx = {1, NULL}
```

which we know works.

File: python/pywatchman/pybser.py [hunk 18]
```diff
@@ -403,12 +411,22 @@ def loads_recursive(self, buf, pos):
                              binascii.hexlify(val_type).decode('ascii'))
 
 def pdu_len(buf):
-    if buf[0:2] != EMPTY_HEADER[0:2]:
+    # The C-extension version of BSER tools does the detection of the protocol
```
Comment (sid0) [inline python/pywatchman/pybser.py]: (This comment isn't necessary -- it's implied that the C and Python versions have the same interface.)

File: python/tests/tests.py [hunk 19]
```diff
@@ -214,13 +214,18 @@ def test_template(self):
 
     def test_pdu_len(self):
         enc = self.bser_mod.dumps(1)
-        self.assertEqual(len(enc), self.bser_mod.pdu_len(enc))
+        # Note: bser.pdu_len returns the PDU length AND the BSER version,
```
Comment (sid0) [inline python/tests/tests.py]: Since this is confusing enough to warrant a comment, maybe rename the function to something more descriptive? :)

File: python/pywatchman/pybser.py [hunk 20]
```diff
@@ -69,6 +69,7 @@ def tobytes(i):
 # our overall length.  To make things simpler, we'll use an
 # int32 for the header
 EMPTY_HEADER = b"\x00\x01\x05\x00\x00\x00\x00"
+EMPTY_HEADER_V2 = b"\x00\x02\x05\x00\x00\x00\x00"
```
Comment (sid0) [inline python/pywatchman/pybser.py]: As a general comment, where are you passing the BSER capabilities down? You probably need another int for that, right? So this should be something more like `"\x00\x02\x05\x00\x00\x00\x00\x05\x00\x00\x00\x00"`, where the first sequence is the capabilities and the second one is the PDU length (or the other way round).

File: python/pywatchman/__init__.py (hunk 11 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: Done.

File: python/pywatchman/__init__.py (hunk 13 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: Done.

File: python/pywatchman/__init__.py (hunk 14 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: Done.

File: python/pywatchman/__init__.py (hunk 15 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: This was a remnant from when I was testing this, so I just removed it.

File: python/pywatchman/__init__.py (hunk 16 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: Done.

File: python/pywatchman/bser.c (hunk 17 above)
Comment (dhruvsinghal) [inline python/pywatchman/bser.c]: Oh this. I'm not entirely sure what mutable does, but I think all I did was just explicitly name the struct members during initialization, since I added a new field to that struct. I'll switch to the old notation {1, 0}.

File: python/pywatchman/pybser.py (hunk 18 above)
Comment (dhruvsinghal) [inline python/pywatchman/pybser.py]: Done.

File: python/tests/tests.py (hunk 19 above)
Comment (dhruvsinghal) [inline python/tests/tests.py]: Done.

**Comment by facebook-github-bot-0** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@dhruvsinghal updated the pull request.

File: python/pywatchman/__init__.py [hunk 21]
```diff
@@ -40,10 +40,10 @@
...
 # Sometimes it's really hard to get Python extensions to compile,
 # so fall back to a pure Python implementation.
-try:
-    from . import bser
-except ImportError:
-    from . import pybser as bser
+#try:
+from . import bser
+#except ImportError:
+#from . import pybser as bser
```
Comment (sid0) [inline python/pywatchman/__init__.py]: You still have this code lying around

**Comment by facebook-github-bot-3** [comment]
@dhruvsinghal updated the pull request.

File: python/pywatchman/__init__.py (hunk 21 above)
Comment (dhruvsinghal) [inline python/pywatchman/__init__.py]: Oops, creeped in again. Fixed.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by sid0** [comment]
Pushed everything up to and including "Fixed tests to work with the new capabilities and BSER v2."

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

File: python/pywatchman/__init__.py [hunk 22]
```diff
@@ -657,9 +657,9 @@ def __init__(self,
             raise WatchmanError('invalid transport %s' % transport)
 
         sendEncoding = str(sendEncoding or os.getenv('WATCHMAN_ENCODING') or
-                           'bser')
+                           'bser-v1')
```
Comment (sid0) [inline python/pywatchman/__init__.py]: Got to continue to call this `'bser'` since it's part of our stable API.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@dhruvsinghal updated the pull request.

File: thirdparty/jansson/value.c [hunk 23]
```diff
@@ -756,6 +760,68 @@ static json_t *json_string_copy(json_t *string)
     return json_string_nocheck(json_string_value(string));
 }
 
+// Given a json array, concat the elements using a delimiter
+w_string_t *w_string_implode(json_t *arr, const char *delim)
```
Comment (sid0) [inline thirdparty/jansson/value.c]: Are you sure you need to do this? Seems like forward declarations + maybe hoisting `json_t` to a new file would work just as well.

I'm concerned because this pokes at internal details like the hash value. It would be nice if we contained all the code that dealt with that in one file.

(BTW if you're doing something non-obvious like this please expand on it in the commit message or in a comment.)

File: string.c [hunk 24]
```diff
@@ -63,18 +62,45 @@ w_string_t *w_string_new_len(const char *str, uint32_t len) {
...
   buf = (char*)(s + 1);
   memcpy(buf, str, len);
   buf[len] = 0;
   s->buf = buf;
+  s->type = type;
 
   return s;
 }
 
+w_string_t *w_string_new_len(const char *str, uint32_t len) {
```
Comment (sid0) [inline string.c]: Separate diff for these, please.

File: string.c [hunk 25]
```diff
@@ -661,7 +661,7 @@ bool w_string_is_null_terminated(w_string_t *str) {
      w_string_is_null_terminated(str->slice));
 }
 
-bool w_string_strlen(w_string_t *str) {
+size_t w_string_strlen(w_string_t *str) {
```
Comment (sid0) [inline string.c]: This should be included in the split out from the previous diff. Remember that each diff must stand independently.

File: cmds/info.c [hunk 26]
```diff
@@ -6,12 +6,13 @@
 static bool query_caps(json_t *response, json_t *result,
     json_t *arr, bool required) {
   size_t i;
+  const char* capname;
```
Comment (sid0) [inline cmds/info.c]: This change seems superfluous.

File: clockspec.c [hunk 27]
```diff
@@ -61,7 +61,8 @@ struct w_clockspec *w_clockspec_parse(json_t *value) {
   if (str[0] == 'n' && str[1] == ':') {
     spec->tag = w_cs_named_cursor;
     // spec owns the ref to the string
-    spec->named_cursor.cursor = w_string_new(str);
+    spec->named_cursor.cursor = json_to_w_string(value);
```
Comment (sid0) [inline clockspec.c]: `json_to_w_string_incref`?

**Comment by sid0** [comment]
The code looks fine overall, thanks. Looks like there are a bunch of ASAN failures indicating memory leaks. I'll let you deal with them.

File: thirdparty/jansson/value.c (hunk 23 above)
Comment (dhruvsinghal) [inline thirdparty/jansson/value.c]: Seems like this functions isn't used anywhere. I'll get rid of it altogether (as discussed offline with @sid0).

File: cmds/info.c (hunk 26 above)
Comment (dhruvsinghal) [inline cmds/info.c]: Fixed.

File: string.c (hunk 24 above)
Comment (dhruvsinghal) [inline string.c]: Done.

File: clockspec.c (hunk 27 above)
Comment (dhruvsinghal) [inline clockspec.c]: Fixed.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@dhruvsinghal updated the pull request.

**Comment by sid0** [comment]
Pushed everything you have so far. Nice work!

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@dhruvsinghal updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@dhruvsinghal updated the pull request.

File: bser.c [hunk 28]
```diff
@@ -353,7 +372,16 @@ int w_bser_dump(const bser_ctx_t* ctx, json_t *json, void *data)
     case JSON_INTEGER:
       return bser_int(ctx, json_integer_value(json), data);
     case JSON_STRING:
-      return bser_bytestring(ctx, json_string_value(json), data);
+      wstr = json_to_w_string(json);
```
Comment (sid0) [inline bser.c]: Wouldn't this affect BSERv1 as well?

**Comment by facebook-github-bot-3** [comment]
@dhruvsinghal updated the pull request - [view changes](https://github.com/facebook/watchman/pull/270/files/0577568c641b6dcceeca6d52f1f82133b99141cd..a4754594e1fa1f749f6d0f02908f7a25a00e95ae)

**Comment by facebook-github-bot-6** [comment]
@dhruvsinghal updated the pull request - [view changes](https://github.com/facebook/watchman/pull/270/files/a4754594e1fa1f749f6d0f02908f7a25a00e95ae..823179f73878df647faa0e79d57e989e3c3da143)

File: bser.c [hunk 29]
```diff
@@ -401,7 +401,7 @@ int w_bser_write_pdu(const uint32_t bser_version,
   }
 
   if (bser_version == 2) {
-    if (bser_int(&ctx, bser_capabilities, data)) {
+    if (dump((const char*) &bser_capabilities, 4, data)) {
```
Comment (sid0) [inline bser.c]: Could you make the 4 a `#define` or similar?

File: python/pywatchman/bser.c [hunk 30]
```diff
@@ -972,9 +972,10 @@ static int pdu_info_helper(PyObject *self, PyObject *args,
   if (bser_version == 2) {
     // Expect an integer telling us what capabilities are supported by the
     // remote server (currently unused).
-    if (!bunser_int(&data, end, &bser_capabilities)) {
+    if (!memcpy(&bser_capabilities, data, 4)) {
```
Comment (sid0) [inline python/pywatchman/bser.c]: Same thing here -- could you make it a `#define`?

File: python/pywatchman/__init__.py [hunk 31]
```diff
@@ -631,15 +631,17 @@ def __init__(self, transport):
...
-        if 'error' in capabilities:
-          raise Exception('Unsupported BSER version')
-
-        if capabilities['capabilities']['bser-v2']:
-            self.bser_version = 2
-            self.bser_capabilities = 0
-        else:
-            self.bser_version = 1
-            self.bser_capabilities = 0
+        # Force the experimental v2 codec to be used. Reactivate fallback
```
Comment (sid0) [inline python/pywatchman/__init__.py]: You probably don't want to have this commit in your final set of changes.

File: bser.c [hunk 32]
```diff
@@ -353,7 +380,16 @@ int w_bser_dump(const bser_ctx_t* ctx, json_t *json, void *data)
     case JSON_INTEGER:
       return bser_int(ctx, json_integer_value(json), data);
     case JSON_STRING:
-      return bser_bytestring(ctx, json_string_value(json), data);
+      wstr = json_to_w_string(json);
+      switch (wstr->type) {
```
Comment (sid0) [inline bser.c]: Add a comment here explaining that `bser_utf8string` is doing to do the version/capabilities check

File: bser.c (hunk 29 above)
Comment (dhruvsinghal) [inline bser.c]: Made it `sizeof(bser_capabilities)`

File: python/pywatchman/bser.c (hunk 30 above)
Comment (dhruvsinghal) [inline python/pywatchman/bser.c]: Made it `sizeof(bser_capabilities)`

File: bser.c (hunk 32 above)
Comment (dhruvsinghal) [inline bser.c]: Done.

**Comment by wez** [comment]
I'm closing this out because it has been idle for a while and has conflicts.  Please open a new pull request rebased off master when you're ready to follow up on this work.
